source/nvmHelper/ns_nvm_helper.c

```diff
@@ -0,0 +1,222 @@
+            // add request to list and handle when existing calls has been handled.
+            ns_list_add_to_end(&ns_nvm_request_list, nvm_request);
+        }
+    } else {
+        ns_nvm_request_t *ns_nvm_request_ptr = ns_nvm_create_request(NULL, NULL, NULL, NULL, NULL, NS_NVM_INIT);
+        if (!ns_nvm_request_ptr) {
+            ns_dyn_mem_free(nvm_request);
+            ns_dyn_mem_free(ns_nvm_request_ptr);
+            return NS_NVM_MEMORY;
+        }
+        ns_nvm_request_ptr->original_request = nvm_request;
+        pnvm_status = platform_nvm_init(ns_nvm_callback_func, ns_nvm_request_ptr);
+        if (pnvm_status != PLATFORM_NVM_OK) {
+            ns_dyn_mem_free(nvm_request);
+            ns_dyn_mem_free(ns_nvm_request_ptr);
+            return NS_NVM_ERROR;
+        }
+        ns_list_init(&ns_nvm_request_list);
+        ns_nvm_initialized = true;
+        ns_nvm_operation_in_progress = true;
+    }
+    return ret;
+}
+
+static ns_nvm_request_t *ns_nvm_create_request(ns_nvm_callback *callback, void *context, const char *key_name, uint8_t *buf, uint16_t *buf_len, uint8_t operation)
+{
+    ns_nvm_request_t *ns_nvm_request_ptr = ns_dyn_mem_alloc(sizeof(ns_nvm_request_t));
+    if (!ns_nvm_request_ptr) {
+        return NULL;
+    }
+    ns_nvm_request_ptr->client_context = context;
+    ns_nvm_request_ptr->callback = callback;
+    ns_nvm_request_ptr->client_key_name = key_name;
+    ns_nvm_request_ptr->operation = operation;
+    ns_nvm_request_ptr->buffer = buf;
+    ns_nvm_request_ptr->buffer_len = buf_len;
+
+    return ns_nvm_request_ptr;
+}
+
+static int ns_nvm_operation_continue(ns_nvm_request_t *request, bool free_request)
+{
+    platform_nvm_status ret = PLATFORM_NVM_OK;
+
+    ns_nvm_operation_in_progress = true;
+    switch(request->operation) {
+        case NS_NVM_KEY_WRITE:
+            request->operation = NS_NVM_KEY_CREATE;
+            ret = platform_nvm_key_create(ns_nvm_callback_func, request->client_key_name, *request->buffer_len, 0, request);
+            break;
+        case NS_NVM_KEY_READ:
+            ret = platform_nvm_read(ns_nvm_callback_func, request->client_key_name, request->buffer, request->buffer_len, request);
+            break;
+        case NS_NVM_KEY_DELETE:
+            ret = platform_nvm_key_delete(ns_nvm_callback_func, request->client_key_name, request);
+            break;
+    }
+
+    if (ret != PLATFORM_NVM_OK) {
+        if (free_request == true) {
+            // free request if requested
+            ns_dyn_mem_free(request);
+        }
+        ns_nvm_operation_in_progress = false;
+        return NS_NVM_ERROR;
+    }
+
+    return NS_NVM_OK;
+}
+
+static void ns_nvm_operation_finalize(ns_nvm_request_t *ns_nvm_request_ptr, int client_retval)
+{
+
+    ns_nvm_request_ptr->callback(client_retval, ns_nvm_request_ptr->client_context);
+    ns_dyn_mem_free(ns_nvm_request_ptr);
+    ns_nvm_operation_in_progress = false;
+
+    while(!ns_list_is_empty(&ns_nvm_request_list)) {
```

Comment:
I'd go with ns_list_foreach_safe() { remove; process; } - "the while !empty get first" idiom is kind of the training wheels version for people who don't trust iterators. The iterator version has the advantage that the loop can have a conditional removal, rather than requiring every element to be removed - you can use the same idiom for more complex cases.